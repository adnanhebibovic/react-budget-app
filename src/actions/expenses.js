import uuid from 'uuid'

export const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date().getTime()
        }
    }
}

export const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}