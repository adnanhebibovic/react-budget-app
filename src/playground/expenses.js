import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
