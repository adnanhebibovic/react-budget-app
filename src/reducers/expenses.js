export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}