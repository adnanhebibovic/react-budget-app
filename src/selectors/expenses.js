export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
