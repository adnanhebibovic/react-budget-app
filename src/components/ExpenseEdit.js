import React from 'react'
import { connect } from 'react-redux'
import { useParams, useHistory } from 'react-router-dom'
import { editExpense, removeExpense } from '../actions/expenses'

import ExpenseForm from './ExpenseForm'

function ExpenseEdit(props) {
    let { id } = useParams()
    let history = useHistory()

    function onSubmit(expense) {
        props.editExpense(id, expense)
        history.push('/')
    }

    function onRemove() {
        props.removeExpense(id)
        history.push('/')
    }

    return (
        <div>
            <h1>Edit Expense</h1>
            <ExpenseForm expense={props.expenses.find((e) => e.id === id)} onSubmit={onSubmit}/>
            <button onClick={onRemove}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expenses: state.expenses
    }
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseEdit)
