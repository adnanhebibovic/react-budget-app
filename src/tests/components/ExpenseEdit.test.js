import React from 'react';
import { Provider } from 'react-redux';
import { create, act } from 'react-test-renderer';
import configureStore from '../../store/expense-store'

import ExpenseEdit from '../../components/ExpenseEdit';

import expenses from '../fixtures/expenses';

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn(),
    useParams: jest.fn()
}));

import { useParams, useHistory } from 'react-router-dom'

let store;
let renderer;

beforeEach(() => {
    useHistory.mockImplementation(() => ({
        push: jest.fn()
    }))

    useParams.mockImplementation(() => ({
        id: expenses[0].id
    }))

    store = configureStore(expenses);

    renderer = create(
        <Provider store={store}>
            <ExpenseEdit />
        </Provider>
    );
});  

test('Should render ExpenseEdit component correctly', () => {
    expect(renderer).toMatchSnapshot();
})

test('Should dispatch an edit action with changed data', () => {
    let expectedState = {
        id: expenses[0].id,
        title: expenses[0].title,
        amount: expenses[0].amount,
        createdAt: expect.any(Number)
    }
    
    let state = store.getState().expenses.filter((e) => e.id == expenses[0].id)
    
    expect(state).toEqual([expectedState])

    const instance = renderer.root;

    act(() => {
        instance.findByProps({type: 'text'}).props.onChange({ target: { value: expenses[1].title } });
    });
    act(() => {
        instance.findByProps({type: 'number'}).props.onChange({ target: { value: expenses[1].amount } });
    });
    act(() => {
        instance.findByType('form').props.onSubmit({ preventDefault: () => jest.fn() });
    });

    expectedState = {
        id: expenses[0].id,
        title: expenses[1].title,
        amount: expenses[1].amount,
        createdAt: expect.any(Number)
    }
    
    state = store.getState().expenses.filter((e) => e.id == expenses[0].id)
    
    expect(state).toEqual([expectedState])
})

test('Should dispatch a remove action with changed data', () => {
    let expectedState = {
        id: expenses[0].id,
        title: expenses[0].title,
        amount: expenses[0].amount,
        createdAt: expect.any(Number)
    }
    
    let state = store.getState().expenses.filter((e) => e.id == expenses[0].id)
    
    expect(state).toEqual([expectedState])

    const instance = renderer.root;

    act(() => {
        instance.findByType('button').props.onClick();
    });

    state = store.getState().expenses.filter((e) => e.id == expenses[0].id)

    expect(state).toEqual([])
})
