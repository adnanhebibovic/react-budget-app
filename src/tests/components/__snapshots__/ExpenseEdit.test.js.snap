// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseEdit component correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <div>
    
    <form
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Title"
        type="text"
        value="first expense"
      />
      <input
        onChange={[Function]}
        placeholder="Amount"
        type="number"
        value={400}
      />
      <input
        type="submit"
        value="Save"
      />
    </form>
  </div>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
